### PR TITLE
Redb slasher backend impl

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

#4424 

## Proposed Changes

add redb implementation to the slasher backend

### CODE REVIEW

**Comment by eserilev** [comment]
hi @michaelsproul 

so I got all the slasher test cases to pass for the redb impl EXCEPT `surrounds_existing_multi_vals_single_chunk`. This test just hangs and never completes. I get a log message saying "test has been running for more than 60 seconds". Not sure whats going on there, will need to spend more time figuring that part out. In general though, everything else seems to be working

The code I wrote probably has a lot of room for improvement. I'm still sort of new to Rust and am having lots of fun with the borrow checker :). In general theres a few things I wanted to mention:

- unlike lmdb and mdbx, redb has a concpt of database tables. so with this redb impl, instead of spinning up 9 separate databases, I created one database and 9 tables. 
- ~~For now I'm opening all the tables in "write" mode. This probably needs to change for functions like `get`, `next_key` etc~~. tables are opened in read mode when relevant
- Theres some general cleanup that needs to happen here. the impl of Database is mostly commented out. I have some functions there I'd like to use, but they dont play nice with the borrow checker. I need to spend some time figuring this out. Once this is resolved the code should be much cleaner
- ~theres a single panic! I need to remove. Theres also a small snippit of useless code there that needs to be changed~
- Do we really need a separate `commit` function? As of now I've included the database commit directly in all of the relevant write functions.

If you get the chance, I could use a review of my changes up to this point. It would be great to get some general feedback so that I can understand if my implementation is at least trending in the right direction

**Comment by eserilev** [comment]
ready for a review

**Comment by eserilev** [comment]
@michaelsproul i'm a bit stuck at the moment and was wondering if you could provide some guidance. Theres a few issues here that I believe make the current slasher db interface incompatible with redb.
 
 I have three separate commit hashes that might help explain the situation.

- Commit hash A `1301e85`
- Commit hash B `41fe56a`
- Commit hash C `3d4fc01`

## Commit hash A

This is a working redb example. I pretty much just did what I needed to do to get things working. Mostly by sprinkling in a few static lifetimes. Not an ideal solution, but I just wanted to get things working so I could look at some metrics.

## Commit hash B

This is roughly how I'd want the redb implementation to look like. When initializing the cursor instance I'm passing in a `WriteTransaction` by reference. This causes some lifetime issues downstream that I just cant seem to resolve

Initially the function that instantiates a new cursor looked something like this

`pub fn cursor<'a>(&'a self, db: &Database) -> Result<Cursor<'a>, Error>`

however this resulted in a lifetime related error message

```
lifetime may not live long enough
consider adding the following bound: `'a: 'env`
requirement occurs because of the type `redb_impl::Cursor<'_>`, which makes the generic argument `'_` invariant
the struct `redb_impl::Cursor<'env>` is invariant over the parameter `'env`
```

following the suggestions the compiler gave me:

`pub fn cursor<'a:'env>(&'a self, db: &Database) -> Result<Cursor<'a>, Error>`

resolves the initial lifetime issues, but now we get some lifetime issues downstream in `database.rs`. There doesn't seem to be a good way to pass the `WriteTransaction` to the `Cursor` with how things are currently defined.

We could start changing the slasher db interface, but thats going to affect the existing slasher db implementations. Probably not an ideal solution...

## Commit hash C

The cursor function now creates a new WriteTransaction instance


``` rust
pub fn cursor<'a>(&'a self, db: &Database) -> Result<Cursor<'a>, Error> {
        let txn = WriteTransaction(self.db.begin_write()?);
        Ok(Cursor {
            current_key: None,
            txn,
            table_name: db.table_name.clone()
        })
    }
```

this closely resembles the lmdb and mdbx implementations. This compiles just fine, but now there exist two `WriteTransaction` instances, one associated to `RwTransaction` and another to `Crursor`. Redb only allows one `WriteTransaction` instance to exist at any one point in time, subsequent instances will be blocked. Under the hood redb uses a mutex to achieve this. Take the following code example

``` rust
write_tx_1 = db.begin_write()?;
write_tx_2 = db.begin_write()?;f

println!("example");
```

`write_tx_1` will get initalized just fine. On the next line, where `write_tx_2` is being initialized, the program will hang. We wont ever get to the print statement until `write_tx_1` is dropped.

Do you think we should explore tweaks to the slasher db interface? Do you have any better ideas?

**Comment by michaelsproul** [comment]
Hey @eserilev thanks for digging deep on this!

I picked up your branch and made a few changes. To address the cursor locking issue I changed the inner `WriteTransaction` to a reference, and then I changed the `Cursor` type so that it has two lifetimes: one for the transaction (`'txn`) and one for the entire database (`'env`). I also clarified some lifetime names in the LMDB/MDBX slasher impls. That commit is here: https://github.com/sigp/lighthouse/pull/4529/commits/d0f983c1b9d4ec0ae5badc9ba8280889c4b7afbb

I also wanted to see what would happen if we pushed `Table` into `Cursor`, which I did here: https://github.com/sigp/lighthouse/pull/4529/commits/90e6bf3911bd8024c49521caee5fdd6f11e8d804. Unfortunately it didn't allow us to eliminate any of the clones (except some trivial ones).

Finally I merged `unstable` back into this PR so we can run CI (and clippy) to catch more unnecessary clones. I'll hand it back over to you now!

File: lighthouse/Cargo.toml [hunk 1]
```diff
@@ -7,7 +7,7 @@ autotests = false
 rust-version = "1.69.0"
 
 [features]
-default = ["slasher-lmdb"]
+default = ["slasher-redb"]
```
Comment (michaelsproul) [inline lighthouse/Cargo.toml]: Before merging we should change the default back to LMDB (for backwards compat).

File: lighthouse/base_db
Comment (michaelsproul) [inline lighthouse/base_db:1]: Should delete this (I'm guessing it's a binary redb file).

File: slasher/Cargo.toml [hunk 2]
```diff
@@ -36,6 +38,8 @@ mdbx = { package = "libmdbx", git = "https://github.com/sigp/libmdbx-rs", tag =
 lmdb-rkv = { git = "https://github.com/sigp/lmdb-rs", rev = "f33845c6469b94265319aac0ed5085597862c27e", optional = true }
 lmdb-rkv-sys = { git = "https://github.com/sigp/lmdb-rs", rev = "f33845c6469b94265319aac0ed5085597862c27e", optional = true }
 
+redb = { version = "1.2.0", optional = true }
```
Comment (michaelsproul) [inline slasher/Cargo.toml]: While resolving the merge from unstable I changed this to point to the release. Hopefully we can stay on upstream redb!

File: slasher/src/database/redb_impl.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+pub struct RwTransaction<'env> {
+    txn: WriteTransaction<'env>,
+    db: &'env redb::Database,
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Cursor<'txn, 'env: 'txn> {
+    txn: &'txn WriteTransaction<'env>,
+    #[derivative(Debug = "ignore")]
```
Comment (michaelsproul) [inline slasher/src/database/redb_impl.rs:37]: we could use this same pattern for `txn` and avoid the wrapper struct

File: slasher/src/database/redb_impl.rs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        &self,
+        db: &Database,
+        key: &K,
+    ) -> Result<Option<Cow<'env, [u8]>>, Error> {
+        let table_definition: TableDefinition<'_, &[u8], &[u8]> =
+            TableDefinition::new(&db.table_name);
+        let table = self.txn.0.open_table(table_definition)?;
+        let result = table.get(key.as_ref().borrow())?;
+        if let Some(access_guard) = result {
+            let value = access_guard.value().to_vec().clone();
```
Comment (michaelsproul) [inline slasher/src/database/redb_impl.rs]: Clippy will probably find this unnecessary clone

**Comment by michaelsproul** [comment]
I was thinking about how to make the cursor more efficient, as I think we're not currently taking full advantage of redb's iteration capabilities. E.g. every time we call `next_key` we construct a new range iterator and call `next` on it. This is going to be `n * log n` = `O(n log n)` to iterate the entire range instead of `O(n)` (iteration of ~2n total BTree nodes).

The cursors are only actually used in a couple of places, and I think we could change the cursor interface in a way that makes it more efficient for redb while maintaining compatibility with the LMDB/MDBX API.

The first usage is here:

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/slasher/src/database.rs#L451-L463

This _just_ uses `last_key` and could be refactored to use a dedicated `last_key` method on `Transaction`. For redb this could work the same as the current `last_key` method, by just constructing an iterator and then immediately calling `next_back`. The put could just use the transaction's `put` instead of a cursor `put`.

The other usage is the complicated one that messes up our use of iterators, because it does deletes using the cursor (which redb doesn't support):

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/slasher/src/database.rs#L695-L716

We could introduce a higher-level abstraction on the `Transaction` interface that looks something like this:

```rust
fn delete_while(&self, database: &Database, f: impl Fn(&[u8]) -> Result<bool, Error>) -> Result<(), Error>;
```

For LMDB/MDBX, the function `f` would replace the `ProposerKey::parse` and `if slot < new_slot` code. For redb we could implement it using [`drain_filter`](https://docs.rs/redb/1.2.0/redb/struct.Table.html#method.drain_filter) which should be more efficient than iterating and deleting keys one at a time.

---

We could also implement this in a follow-up PR, as I think a 4x slowdown compared to LMDB is acceptable for an initial merge to `unstable`.

PS: Looks like the tests are failing on CI due to some lifetime tweaks. Let me know if you need a hand with them.

**Comment by eserilev** [comment]
Thanks for the great feedback! I'm not in a big rush to get this merged, so I think it makes sense to spend time optimizing things here a bit.

> This just uses last_key and could be refactored to use a dedicated last_key method on Transaction. For redb this could work the same as the current last_key method, by just constructing an iterator and then immediately calling next_back. The put could just use the transaction's put instead of a cursor put.

I'm looking at that bit of code and trying to understand the implications of moving `last_key` from `Cursor`  to `Transaction` .  Are you suggesting this just as a way to move away from using `Cursor`, or is there some additional optimization being done that I'm not understanding?

Your second suggestion makes sense to me.

> PS: Looks like the tests are failing on CI due to some lifetime tweaks. Let me know if you need a hand with them.

I think my most recent commit resolves the lifetime issues for lmdb. I believe a similar fix is needed for mdbx. I'm facing a bit of difficulty building when mdbx is the selected db 

```
Thread 'main' panicked at '"MDBX_version_info_struct_(unnamed_at_/Users/eitan/_cargo/git/checkouts/libmdbx-rs-c1b523f5b64ff08c/096da80/mdbx-sys/libmdbx/mdbx_h_611_3)" is not a valid Ident', /Users/eitan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.67/src/fallback.rs:774:9
```

Might be an issue specifically with my branch? I'll try building on unstable with mdbx as the default slasher db and see what happens. If I cant resolve, I'll probably ask for help in discord.

**Comment by michaelsproul** [comment]
@eserilev Yeah MDBX can be a bit annoying to build (see #4280). I think that issue is caused by the libclang version. On Ubuntu 22.04 I've got these packages installed:

```
$ apt list --installed | grep clang
clang-13/lunar-updates,lunar-security,now 1:13.0.1-11ubuntu14.1 amd64 [installed,automatic]
clang-14/lunar-updates,lunar-security,now 1:14.0.6-12ubuntu0.23.04.1 amd64 [installed,automatic]
clang-15/lunar-updates,lunar-security,now 1:15.0.7-3ubuntu0.23.04.1 amd64 [installed,automatic]
clang/lunar,now 1:15.0-56~exp2 amd64 [installed]
libclang-common-13-dev/lunar-updates,lunar-security,now 1:13.0.1-11ubuntu14.1 amd64 [installed,automatic]
libclang-common-14-dev/lunar-updates,lunar-updates,lunar-security,lunar-security,now 1:14.0.6-12ubuntu0.23.04.1 all [installed,automatic]
libclang-common-15-dev/lunar-updates,lunar-updates,lunar-security,lunar-security,now 1:15.0.7-3ubuntu0.23.04.1 all [installed,automatic]
libclang-cpp13/lunar-updates,lunar-security,now 1:13.0.1-11ubuntu14.1 amd64 [installed,automatic]
libclang-cpp14/lunar-updates,lunar-security,now 1:14.0.6-12ubuntu0.23.04.1 amd64 [installed,automatic]
libclang-cpp15/lunar-updates,lunar-security,now 1:15.0.7-3ubuntu0.23.04.1 amd64 [installed,automatic]
libclang-rt-14-dev/lunar-updates,lunar-security,now 1:14.0.6-12ubuntu0.23.04.1 amd64 [installed,automatic]
libclang-rt-15-dev/lunar-updates,lunar-security,now 1:15.0.7-3ubuntu0.23.04.1 amd64 [installed,automatic]
libclang1-13/lunar-updates,lunar-security,now 1:13.0.1-11ubuntu14.1 amd64 [installed,automatic]
libclang1-14/lunar-updates,lunar-security,now 1:14.0.6-12ubuntu0.23.04.1 amd64 [installed,automatic]
libclang1-15/lunar-updates,lunar-security,now 1:15.0.7-3ubuntu0.23.04.1 amd64 [installed,automatic]
```

If you can't get it working I can fix it up.

I also think your fix for LMDB is a little bit inefficient in that it uses `.to_vec()`. We can probably tweak the lifetimes on the `Cow` so that only `redb` uses `to_vec` and the other backends can continue to return references.

> I'm looking at that bit of code and trying to understand the implications of moving last_key from Cursor to Transaction . Are you suggesting this just as a way to move away from using Cursor, or is there some additional optimization being done that I'm not understanding?

Yeah it's just a way to get rid of cursor. Alternatively we could keep cursor, with just that `last_key` method, but it seems a bit unnecessary

**Comment by eserilev** [comment]
> I also think your fix for LMDB is a little bit inefficient in that it uses .to_vec(). We can probably tweak the lifetimes on the Cow so that only redb uses to_vec and the other backends can continue to return references.

yep, you're right. I tweaked the Cow lifetime so that we can continue returning references for lmdb/mdbx

**Comment by eserilev** [comment]
here are some updated metrics after applying the optimizations we talked about
https://snapshots.raintank.io/dashboard/snapshot/DhQWwC4l7Cel2nBwvnI27bAK76qBC3LI

Actually seems to be a bit slower than before, but I think the reason for that may be related to the changes done to the shared execution node. Might be because i'm not ssh-ing into the exec client anymore, or maybe its related to jwt auth? 
I actually had to set `execution-timeout-multiplier` to 3 because my requests were timing out with the default settings (I set `execution-jwt-id` as well). I'm going to run an earlier redb commit hash to verify that the performance degradation is actually just due to execution node slowdown and not because of my changes. 

I pushed up the optimization changes, currently its only been implemented for redb. Havent gotten around to updating lmdb and mdbx yet

Also, my optimization changes were only applied to `prune_proposers`. `prune_indexed_attestations` still contains the same inefficiencies (for the redb impl)

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/slasher/src/database.rs#L740-L760

Because we're mutating within the loop, the function `f` in `delete_while` would need to be mutable. `drain_filter` doesn't allow for mutability within `predicate`. However `drain_filter` does return an iterator containing the deleted key/value pairs, maybe it makes sense for `delete_while` to return a list of key/value pairs. Will explore that option

EDIT:

redb metrics from an old commit hash https://snapshots.raintank.io/dashboard/snapshot/AKQGF4JaAf3FmIS8ISOdntHATjZt56nu

looks like the slowdown wasn't related to any of my recent code changes

**Comment by michaelsproul** [comment]
Nice!

Yeah I think having `delete_while` return the list of deleted key/value pairs sounds good.

Another more hacky approach would be to use an `Arc<Mutex<Vec<_>>>` and do `vec.lock().push(..)` from within the immutable `f`.

I'm hopeful that changing the cursor usage in attestation pruning will give us the speed-up we're looking for. It's true there are a lot more attestations than blocks, so operations involving attestations _should_ comprise more of the slasher's run time 🤞

**Comment by eserilev** [comment]
These are the updated metrics w/ optimizations implemented for `prune_proposer` and `prune_Indexed_attestation`

https://snapshots.raintank.io/dashboard/snapshot/bnIAVIgu0Hl4FY0VbdBdlG5i4tinFhIf

I dont think I'm seeing a big difference here in terms of performance, potentially maybe a slight improvement. One additional optimization we could make to redb is performing compaction on some interval. This could potentially result in less disk usage compared to LMDB. Not sure where we could sneak in calls to compact the database but it might be worth exploring

File: slasher/src/database.rs [hunk 5]
```diff
@@ -727,37 +722,29 @@ impl<E: EthSpec> SlasherDB<E> {
...
-                break;
-            }
-        }
+        let indexed_attestation_ids: Vec<IndexedAttestationId> = cursor
+            .delete_while(should_delete)?
+            .into_iter()
+            .map(|value| {
+                IndexedAttestationId::new(IndexedAttestationId::parse(value).unwrap_or_default())
+            })
+            .collect();
```
Comment (michaelsproul) [inline slasher/src/database.rs]: we could collect into a `Result<Vec<_>, _>>` here to handle the errors from `::parse`

File: slasher/src/database/redb_impl.rs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+pub struct Cursor<'env> {
+    #[derivative(Debug = "ignore")]
+    txn: &'env redb::WriteTransaction,
+    db: &'env Database<'env>,
+    current_key: Option<Cow<'env, [u8]>>,
+}
+
+impl Environment {
+    pub fn new(config: &Config) -> Result<Environment, Error> {
+        let db_path = match config.database_path.join(BASE_DB).as_path().to_str() {
```
Comment (michaelsproul) [inline slasher/src/database/redb_impl.rs]: do we definitely need to join the `BASE_DB` here? I think the caller takes care of that already when constructing the `config.database_path`

File: lighthouse/tests/beacon_node.rs [hunk 7]
```diff
@@ -2187,6 +2187,8 @@ fn slasher_broadcast_flag_false() {
             assert!(!slasher_config.broadcast);
         });
 }
+
+#[cfg(all(feature = "lmdb", not(any(feature = "lmdb", feature = "mdbx"))))]
```
Comment (michaelsproul) [inline lighthouse/tests/beacon_node.rs]: this `cfg` always evaluates to `false`, it's checking for "LMDB and not LMDB"

Comment (michaelsproul) [reply lighthouse/tests/beacon_node.rs]: I think this test should be able to pass as long as LMDB is enabled

**Comment by michaelsproul** [comment]
I just pushed a minor optimisation that avoids collecting into an intermediate Vec: https://github.com/sigp/lighthouse/pull/4529/commits/6b0a81ff26925836cadda9d36bc402c5da86bb77

**Comment by michaelsproul** [comment]
Also just back-merged unstable so the version reads 5.2

File: lighthouse/tests/beacon_node.rs (hunk 7 above)
Comment (eserilev) [reply lighthouse/tests/beacon_node.rs]: re-enabled test in [df800b0](https://github.com/sigp/lighthouse/pull/4529/commits/df800b0bbfc90b6daed306ff379e867cb09db1cc)

**Review by michaelsproul — APPROVED** [review]
Just pushed one change to tidy up the DB filename, which we would have trouble changing later.

**Comment by michaelsproul** [comment]
@mergifyio queue

**Pull request merged**